lib/application_group.ts

```diff
@@ -0,0 +1,224 @@
+import { RawStorageKey } from '@Lib/storage_keys';
+import { removeFromArray, findInArray } from '@Lib/utils';
+import { UuidString, DeinitSource } from './types';
+import { SNApplication } from './application';
+import { PureService } from '@Services/pure_service';
+import { DeviceInterface } from '@Lib/device_interface';
+import { Uuid } from '@Lib/uuid';
+
+export type ApplicationDescriptor = {
+  identifier: string | UuidString,
+  // userUuid?: UuidString,
+  label: string,
+  /** Whether the application is the primary user-facing selected application */
+  primary: boolean,
+}
+
+export type DescriptorRecord = Record<string, ApplicationDescriptor>
+
+type AppGroupCallback = {
+  applicationCreator: (
+    descriptor: ApplicationDescriptor,
+    deviceInterface: DeviceInterface
+  ) => SNApplication
+}
+
+type AppGroupChangeCallback = () => void
+
+export class SNApplicationGroup extends PureService {
+  public primaryApplication?: SNApplication
+  private descriptorRecord!: DescriptorRecord
+  private changeObservers: AppGroupChangeCallback[] = []
+  callback!: AppGroupCallback
+  private applications: SNApplication[] = []
+
+  constructor(public deviceInterface: DeviceInterface) {
+    super();
+  }
+
+  deinit() {
+    super.deinit();
+    this.deviceInterface.deinit();
+    (this.deviceInterface as any) = undefined;
+  }
+
+  public async initialize(callback: AppGroupCallback) {
+    this.callback = callback;
+    this.descriptorRecord = (
+      await this.deviceInterface!.getJsonParsedRawStorageValue(RawStorageKey.DescriptorRecord)
+    ) as DescriptorRecord;
```

Comment:
Can we trust this value enough to cast it? What would happen if it wasn't formatted in the way we expect?

Reply:
Yup we are the same class that does the persisting.